### PR TITLE
feat: remote pinning settings page

### PR DESCRIPTION
**Pull request by rafaelramalho19**
## Description

- Unstubs UI Elements to use Remote Pinning API (https://github.com/ipfs/js-ipfs/pull/3293)
- The CI is probably going to fail until we have a go-ipfs@8+ release and bump this PR to use it

## Testing

To test this locally, follow these steps: 
- Checkout js-ipfs into local dir and switch to @Gozala branch (feat/pin-remote)
- Go to js-ipfs/package/ipfs-http-client and try to build it via `npm ci`.
- Go back to ipfs-webui dir, and execute `connect-deps link /path/to/js-ipfs/packages/ipfs-http-client` and then `connect-deps connect`
if successful, you should be able to run `npm i && npm start` and have getIpfs().pin.remote

### CODE REVIEW

**Review by jessicaschilling — CHANGES_REQUESTED** [review]
LGTM (based on the demo video - let me know if I need to dig deeper) with only one request: that the colors allocated to services' icons remain consistent. Thank you!

**Review by lidel — CHANGES_REQUESTED** [review]
Thank you @rafaelramalho19! 

I made a quick pass and wrote down some nits, so we don't miss them:

- :broken_heart: When I open _Settings_ for the first time it does not show me my pre-existing services.
   - I believe we should execute `pin.remote.service.ls` on initial page load as a quick test of `pin remote` functionality: 
     - if it works and returns a list (could be empty), trigger rendering of  the list
     - if it errors (eg. because backend is go-ipfs older than 0.7.0) then disable "Add service" button and replace onboarding text with some error message. 
       - @jessicaschilling we probably should prepare new onboarding text displayed above the list of pinning services when `pin remote` commands are present. Right now the user is presented with:
        https://github.com/ipfs-shipyard/ipfs-webui/blob/53ebbce1ebae2ea0b483eebf45d71a81c068ccf8/public/locales/en/settings.json#L8
        We  could keep this for situation when `pin.remote` is not available, but we need new one for when it is present.
- :green_heart: Adding a new service works as expected and after that I see newly added one + pre-existing ones I had already defined via CLI: 
  Screenshot_2021-01-25 Settings IPFS
- :thinking: e2e tests fail with 0.8.0-rc1  but I can take a look at them after we fix regular `npm test` which fail with `TypeError: TextDecoder is not a constructor` (which seems to be due to some changes in `ipfs-http-client`) – don't worry about those yet, we will circle back after wiring of Files screen is finished  
- :thinking: Clicking on "Set pinning" on Files screen fails with [`TypeError: remoteServices is not iterable`](https://user-images.githubusercontent.com/157609/105653701-bcd6fa00-5ebc-11eb-9280-43c16813f885.png)  
  (not a big deal, but if  you will work on Files in separate PR, then would be cool to stub it here for now, so it does not break, and we can emerge this PR independently)

**Comment by jessicaschilling** [comment]
We have that text already in an earlier branch ... somewhere. I'll dig it out on Monday.

File: src/components/pinning-manager/PinningManager.js
Comment (rafaelramalho19) [reply src/components/pinning-manager/PinningManager.js:37]: I Haven't seen it fail yet, it's just defensive, but good suggestion :)

File: src/bundles/pinning.js
Comment (rafaelramalho19) [reply src/bundles/pinning.js]: Good idea, moved it to constants/pinning.js 👍

**Comment by lidel** [comment]
@rafaelramalho19  I'm seeing the local pinning, no "add service" button + this error in the console:
> 

Is it something on my end? 
FYSA I'm building with the latest version of Irakli's PR, perhaps you had an older version checked out?

**Comment by rafaelramalho19** [comment]
> @rafaelramalho19 I'm seeing the local pinning, no "add service" button + this error in the console:
> 
> > 
> 
> Is it something on my end?
> FYSA I'm building with the latest version of Irakli's PR, perhaps you had an older version checked out?
> 
> ps. ignore E2E tests for now, I want us to fix them in #1655 first

Happens to me too if I update the code with Irakli's latest 3 commits. I'll look into it

**Review by lidel — CHANGES_REQUESTED** [review]
We merged latest http-client with https://github.com/ipfs-shipyard/ipfs-webui/pull/1655 so connect-deps should no longer be necessary.

@rafaelramalho19 mind rebasing this PR?

**Comment by lidel** [comment]
I've merged fixes from master and switched Pinata endpoint to production – seems there were no bugs, things work as expected.  

@jessicaschilling mind trying to build this branch locally when you have some spare time?  regular `npm ci && npm start` should just work (no need for `connect-deps` anymore)

I think we could hide Auto Upload column for now and merge it to unblock https://github.com/ipfs-shipyard/ipfs-webui/pull/1721   
(MFS autopinning can be added in a separate PR later).

**Comment by jessicaschilling** [comment]
@lidel looks good to me - I don't have a Pinata account to test on, but shoot me details OOB if you want me to specifically test. Agree that removing the auto upload column to push the feature out faster is a good move.

**Comment by lidel** [comment]
@rafaelramalho19 quick notes on remaining work here:

- [ ] Files screen still breaks when one of your pinning services is offline ([screenshot](https://user-images.githubusercontent.com/157609/111212757-4c9d4880-85d0-11eb-8c84-6c186c4d0dae.png))
- As you have bandwidth now, please wire up Auto Upload column to toggle [`Enable`](https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#pinningremoteservices-policiesmfsenabled) flag in MFS Pining policy
  - this can be done via `ipfs.config`, on the CLI it works like this:
    ```
    ipfs config --json Pinning.RemoteServices.your-service.Policies.MFS.Enable false
    ```
- [ ]  On my test node Settings page fails to calculate their size due to [goPinsSizeGet error](https://user-images.githubusercontent.com/157609/111210888-165ec980-85ce-11eb-95c0-6d86fbc59e37.png). I am not sure if error is due to that, or something else. 
  - I have 1k of local pins and multiple remote services – unlikely the cause, but mentioning for the record.
  - Some of remote services are broken (offline) – perhaps this is the cause?
- [ ]  (Assuming previous bug is fixed) Calling `files.stat` for each CID in `goPinsSizeGet`  is too slow to be execute every time Settings screen is loaded. some CIDs may represent huge DAGs and calculating size of each takes time
  - Try adding a localStorage-backed cache that memoizes the `cumulativeSize`  of each CID, so `files.stat` is called only once. Maybe that is enough.  
- [ ]  Fill a separate issue to revert d20e4e30  and investigate why JS fails  (so we dont forget)

**Comment by rafaelramalho19** [comment]
> On my test node Settings page fails to calculate their size due to goPinsSizeGet error. I am not sure if error is due to that, or something else.

This was due to ipfs not being ready at the time of trying to fetch the services, I changed how that works, see if it's fixed now 🙏

**Comment by lidel** [comment]
@rafaelramalho19  quick feedback/clarifications after your force-push on Friday:
- Files screen calling  `pin.remote` against service which is offline can be reproduced in you set up something like [this](https://user-images.githubusercontent.com/157609/111975234-a3d77780-8b00-11eb-8b95-f2adf466829c.png) (where  remote pinning service endpoint is `http://127.0.0.1:5000` and I have nothing running on that port)
- Problem with the   [lack of error handling in `goPinsSizeGet` ](https://user-images.githubusercontent.com/157609/111980682-f156e300-8b06-11eb-9f81-5ee7f5f5d57f.png) can be reproduced with `QmYw8okuEpxCPhfX3CKbF8mtpmyHPHP9VC3VNhUepytcSh` (triggers EOF error):
  ```
  curl -X POST 'http://127.0.0.1:5001/api/v0/files/stat?arg=%2Fipfs%2FQmYw8okuEpxCPhfX3CKbF8mtpmyHPHP9VC3VNhUepytcSh'
  ```

**Comment by rafaelramalho19** [comment]
> * Files screen calling  `pin.remote` against service which is offline can be reproduced in you set up something like [this](https://user-images.githubusercontent.com/157609/111975234-a3d77780-8b00-11eb-8b95-f2adf466829c.png) (where  remote pinning service endpoint is `http://127.0.0.1:5000` and I have nothing running on that port)

I tested that out and for me it fails gracefully, since I just ignore dev errors as a usual user, which is the use-case for all users in IPFS unless they run in dev mode: 
image

**Comment by rafaelramalho19** [comment]
> * Problem with the   [lack of error handling in `goPinsSizeGet` ](https://user-images.githubusercontent.com/157609/111980682-f156e300-8b06-11eb-9f81-5ee7f5f5d57f.png) can be reproduced with `QmYw8okuEpxCPhfX3CKbF8mtpmyHPHP9VC3VNhUepytcSh` (triggers EOF error):
>   ```
>   curl -X POST 'http://127.0.0.1:5001/api/v0/files/stat?arg=%2Fipfs%2FQmYw8okuEpxCPhfX3CKbF8mtpmyHPHP9VC3VNhUepytcSh'
>   ```

For me it happens in master too (without the pinning services integration) 
image
So are we looking at a bigger issue here? The lack of error handling upon all EOFs errors?

**Comment by lidel** [comment]
@rafaelramalho19 I believe master branch has the "stage 1"  – "Pinning Services" section only with local pins, and calculating the size of local pins via `ipfs.files.stat` call, so it also fails. 

We need to either fix error handling or disable "size" column for now.

It is possible the other error is only a side effect of this.

**Comment by lidel** [comment]
Filled https://github.com/ipfs/go-ipfs/issues/7999 for a variant of the `unexpected EOF` error.
Still, we need to add error handling to webui (if user pinned or imported  `QmYw8okuEpxCPhfX3CKbF8mtpmyHPHP9VC3VNhUepytcSh`  to MFS, calculating size should not break the app)

**Comment by lidel** [comment]
@rafaelramalho19  good news, I was able to fix those random errors: :partying_face: 

- [x] Made size calculations immune to unexpected error + disabled pin size calculation in https://github.com/ipfs-shipyard/ipfs-webui/pull/1713/commits/b0c18d771e9e8b690e30d7b895c265379134545d
- [x] https://github.com/ipfs-shipyard/ipfs-webui/pull/1713/commits/5d6a1b4e9266f4f187826b3792057b90d66097f0 fixed a bug when if one of services returned error for `pin.remote.ls`  (that stopped fetching remote pins from remaining services) – this was specific to Firefox and more than 4 services, so hard to replicate.

Remaining actionable UX pains which you might be able to pick up:

- [ ] **disable offline services in pinning modal**
  I think we may avoid  all sorts of problems if  "Set pinning" modal executes `ipfs pin remote service ls --stat`  check (just like we do on _Settings_) and disables services that are not online, so user can't pin to service that is offline and avoids error.

- [ ] **delete modal does not unpin**
  We track this in https://github.com/ipfs-shipyard/ipfs-webui/pull/1644, is that PR waiting for this one to be merged, or is it other way around?

Things that I am unsure how to tackle (would appreciate thoughts/ideas):

- [ ] **can't tell when a file is pinned indirectly** (when file is not pinned, but its parent dir is remotely pinned)
  This is especially problematic when MFS autopinning is enabled. In that scenario, everything is pinned indirectly via MFS root CID, but that is not reflected in the Files user interface in any way.
  (should we display a faded cloud icon? new icon for indirect pins?)

- [ ] **fetching all remote pins will  become very expensive over time**
  Right now we fetch all remote pins via `pin.remote.ls` once, and then check cache while browsing Files.
  What if user has 1k of remote pins, or more? It will take multiple seconds/minutes to before webui can show remote pin status for the first time, and the entire list will have to be re-fetched after page reload.
  Perhaps we could fetch pin status progressively, only for CIDs in current directory + every parent directory?
  API allows fetching status for up to 10 cids per request:
  `pin.remote.ls({service: 'foo', cid: [cid1, cid2, ... cid10]})`

Text updates (ok to ignore for now, writing down so we don't forget before merge):

- [ ] update  Auto Upload things
  - tweak menu item case to match other labels
  - update [explainer in modal](https://github.com/ipfs-shipyard/ipfs-webui/blob/5d6a1b4e9266f4f187826b3792057b90d66097f0/public/locales/en/settings.json#L45)
  - replace "Toggle" with "Enable / Disable" in modal title and button
  - replace "N/A" in Auto Upload column with "Off"  (and add it to locale file)
- [ ] update explainer on  Settings screen
  - for some reason `noServicesDescription` is displayed even when eservices are defined
  - [update explainer from stage 1](https://github.com/ipfs-shipyard/ipfs-webui/blob/5d6a1b4e9266f4f187826b3792057b90d66097f0/public/locales/en/settings.json#L9)

**Comment by rafaelramalho19** [comment]
>delete modal does not unpin locally by default
> We track this in #1644, is that PR waiting for this one to be merged, or is it other way around?

The modal PR is waiting for this, since it's not even functional without the proper API for unpinning files

> disable offline services in pinning modal

Unsure what you mean about this, since everytime we open the pinning modal we fetch the remote pin services again (`doFetchPinningServices`) which updates the list, or am I missing something?

**Comment by lidel** [comment]
What I meant is that we should disable checkbox next to services that are known to be offline. 
(And fade incactive ones out a bit, so user can tell those are not available for pinning)

**Comment by lidel** [comment]
Pushed some changes:

- filled missing labels with basic explainer (canbe improved in separate PRs)
- disabled pinning to services that are offline + added visual indication when there is a problem with service

Screenshot_2021-03-25  animals pets Files IPFS
Screenshot_2021-03-25 Settings IPFS
Screenshot_2021-03-25 Settings IPFS(1)

File: public/locales/en/settings.json
Comment (jessicaschilling) [inline public/locales/en/settings.json:45]:
```suggestion
    "title": "Change upload policy for {name}",
```

File: public/locales/en/settings.json
Comment (jessicaschilling) [inline public/locales/en/settings.json:46]:
```suggestion
    "description": "When enabled, auto upload will periodically pin the root of your Files directory, ensuring everything is pinned remotely and available to other peers even when your local IPFS node goes offline. If you disable it, you can still pin each file manually using the \"Set pinning\" option on the Files screen."
```

**Review by jessicaschilling — COMMENTED** [review]
Swooped in with some small text edits ... hope that's OK 😊

**Comment by lidel** [comment]
Fixed icon/link support for remote service templates in GUI.
If name includes template name (for now we only have "pinata" one), then icon + link to docs from template are used on Settings:

> 

.. and Files:

> 


TODO (can be filled as  separate issue):

- [ ] store only CID in template, and use same gateway for loading the icon as FilePreviews on Files screen

File: public/locales/en/settings.json
Comment (jessicaschilling) [inline public/locales/en/settings.json:11]:
```suggestion
    "addAutoUpload": "Enable auto upload",
```

**Review by lidel — APPROVED** [review]
Thank you @rafaelramalho19 for everything that went into this one :heart: 
As this PR was initially about Settings, I believe it is functionally complete and good to merge to unblock https://github.com/ipfs-shipyard/ipfs-webui/pull/1644.  


I extracted the remaining work into separate issues and put them in [milestone v2.12](https://github.com/ipfs/ipfs-webui/milestone/4).
@rafaelramalho19 when you feel better, feel free to pic tasks from mentioned milestone, starting with #1644. Remember to self-assign, so we don't duplicate work.

We are nearly there :+1:

**Pull request merged**
